clients/drcachesim/tracer/raw2trace.h

```diff
@@ -1034,46 +1034,62 @@ template <typename T> class trace_converter_t {
+    handle_kernel_interrupt_and_markers(void *tls, INOUT trace_entry_t **buf_in,
+                                        uint64_t cur_modoffs, int instr_length,
+                                        OUT bool *interrupted)
     {
         // To avoid having to backtrack later, we read ahead to ensure we insert
         // an interrupt at the right place between memrefs or between instructions.
         *interrupted = false;
-        const offline_entry_t *in_entry = impl()->get_next_entry(tls);
-        if (in_entry == nullptr)
-            return "";
-        if (in_entry->extended.type == OFFLINE_TYPE_EXTENDED &&
-            in_entry->extended.ext == OFFLINE_EXT_TYPE_MARKER &&
-            in_entry->extended.valueB == TRACE_MARKER_TYPE_KERNEL_EVENT) {
-            // A signal/exception marker in the next entry could be at any point
-            // among non-memref instrs, or it could be after this bb.
-            // We check the stored offset.
-            uint64_t int_modoffs = (uint64_t)in_entry->extended.valueA;
-            impl()->log(4,
-                        "Checking whether reached signal/exception +" PIFX
-                        " vs cur +" PIFX "\n",
-                        int_modoffs, cur_modoffs);
-            // Because we increment the instr fetch first, the signal modoffs may be
-            // less than the current for a memref fault.
-            if (int_modoffs == cur_modoffs || int_modoffs + instr_length == cur_modoffs) {
-                impl()->log(4, "Signal/exception interrupted the bb @ +" PIFX "\n",
-                            int_modoffs);
+        bool append = false;
+        do {
+            const offline_entry_t *in_entry = impl()->get_next_entry(tls);
+            if (in_entry == nullptr)
+                return "";
+            if (in_entry->extended.type != OFFLINE_TYPE_EXTENDED ||
+                in_entry->extended.ext != OFFLINE_EXT_TYPE_MARKER) {
+                append = false;
+            } else if (in_entry->extended.valueB == TRACE_MARKER_TYPE_KERNEL_EVENT) {
+                // A signal/exception marker in the next entry could be at any point
+                // among non-memref instrs, or it could be after this bb.
+                // We check the stored offset.
+                uint64_t int_modoffs = (uint64_t)in_entry->extended.valueA;
+                impl()->log(4,
+                            "Checking whether reached signal/exception +" PIFX
+                            " vs cur +" PIFX "\n",
+                            int_modoffs, cur_modoffs);
+                // Because we increment the instr fetch first, the signal modoffs may be
+                // less than the current for a memref fault.
+                if (int_modoffs == cur_modoffs ||
+                    int_modoffs + instr_length == cur_modoffs) {
+                    impl()->log(4, "Signal/exception interrupted the bb @ +" PIFX "\n",
+                                int_modoffs);
+                    append = true;
```

Comment:
I can't tell for sure because I don't think I have fully groked the logic here, but it seems weird to me that we are setting "append" in this if branch but not resetting it to false in the case that there wasn't an interruption. Instead, "append" keeps the same value as before (false on first iteration, true on subsequent iterations), which seems incorrect or at the very least tricky.

I know there's already a bunch of comments here, and maybe I accidentally missed this info in them, but if this is intentional can you add some explicit mention of this into a comment?

Reply:
Actually that is a bug: for a kernel interrupt whose PC does not match.  Hmm, I would have hoped the drcachesim.invariants test would catch that -- but I guess it relies on actual signal arrivals and is not guaranteed to exercise it.  IIRC it was not easy to make a targeted test there.  In any case it is now fixed.  I also added a comment clarifying why it's ok to exit the loop with other markers after a non-matching interrupt marker.